src/raftstore/store/worker/pd.rs

```diff
@@ -14,78 +14,143 @@
-            }
             Task::AskSplit { ref region, ref split_key, .. } => {
                 write!(f,
                        "ask split region {} with key {}",
                        region.get_id(),
                        escape(&split_key))
             }
-            Task::Heartbeat { ref store } => write!(f, "heartbeat for store {}", store.get_id()),
+            Task::Heartbeat { ref region, .. } => {
+                write!(f, "heartbeat for region {}", region.get_id())
+            }
         }
     }
 }
 
 pub struct Runner<T: PdClient> {
     pd_client: Arc<T>,
+    ch: SendCh,
 }
 
 impl<T: PdClient> Runner<T> {
-    pub fn new(pd_client: Arc<T>) -> Runner<T> {
-        Runner { pd_client: pd_client }
+    pub fn new(pd_client: Arc<T>, ch: SendCh) -> Runner<T> {
+        Runner {
+            pd_client: pd_client,
+            ch: ch,
+        }
+    }
+
+    fn send_admin_request(&self,
+                          mut region: metapb::Region,
+                          peer: metapb::Peer,
+                          request: AdminRequest) {
+        let region_id = region.get_id();
+        let admin_type = request.get_cmd_type();
+
+        let mut req = RaftCmdRequest::new();
+        req.mut_header().set_region_id(region_id);
+        req.mut_header().set_region_epoch(region.take_region_epoch());
+        req.mut_header().set_peer(peer);
+        req.mut_header().set_uuid(Uuid::new_v4().as_bytes().to_vec());
+
+        req.set_admin_request(request);
+
+        let cb = Box::new(move |_: RaftCmdResponse| -> Result<()> { Ok(()) });
+
+        if let Err(e) = self.ch.send(Msg::RaftCmd {
+            request: req,
+            callback: cb,
+        }) {
+            error!("send {:?} request to region {} err {:?}",
+                   admin_type,
+                   region_id,
+                   e);
+        }
     }
 }
 
 impl<T: PdClient> Runnable<Task> for Runner<T> {
     fn run(&mut self, task: Task) {
-        info!("executing task {}", task);
+        debug!("executing task {}", task);
 
-        let res = match task {
-            Task::AskChangePeer { region, peer, .. } => {
-                // TODO: We may add change_type in pd protocol later.
-                self.pd_client.ask_change_peer(region, peer)
-            }
+        match task {
             Task::AskSplit { region, split_key, peer } => {
-                self.pd_client.ask_split(region, &split_key, peer)
+                if let Ok(mut resp) = self.pd_client.ask_split(region.clone()) {
+                    info!("try to split with new region id {} for region {:?}",
+                          resp.get_new_region_id(),
+                          region);
+                    let req = new_split_region_request(split_key,
+                                                       resp.get_new_region_id(),
+                                                       resp.take_new_peer_ids());
+                    self.send_admin_request(region, peer, req);
```

Comment:
no matter now,  we use one connection to communicate pd synchronously, and it is easy to change asynchronous way after we introduces HTTP. And the new mechanism doesn't need immediacy, so it is ok here. 

Btw, ask split is very fast in pd now and even faster than heartbeat.